doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -525,25 +525,36 @@ This section describes some special purpose tactics to work with
 
 .. tacn:: discriminate {? @induction_arg }
 
-   Proves any goal for which a hypothesis in the form :n:`@term__1 = @term__2`
-   states an impossible structural equality for an inductive type.
-   If :n:`@induction_arg` is not given, it checks all the hypotheses
+   Proves goals for which a hypothesis in the form :n:`@term__1 = @term__2`
+   has inconsistent constructors between the two sides of
+   the equality.
+   The tactic relies on the fact that constructors of inductive types are injective
+   and disjoint, i.e. if `C1` and `C2` are distinct constructors of an inductive type then
+   :n:`C1 @term__1 = C1 @term__2` implies that :n:`@term__1 = @term__2` (injectivity)
+   and :n:`C1 @term__1 = C2 @term__2` is a contradiction (disjointedness).
+   For example, :g:`S (S O) = S O` is a contradiction: while
+   the outermost constructors are both `S`, the next ones differ (`S` versus `O`).
+
+   The tactic traverses the normal forms of :n:`@term__1` and :n:`@term__2`,
+   looking for subterms placed in the same positions and whose
+   head symbols are different constructors. If such subterms are present, the
+   equality is impossible and the current goal is completed.
+   Otherwise the tactic fails.  Note that opaque constants are not expanded by
+   δ reductions while computing the head normal form.
+
+   Note that :n:`discriminate` doesn't handle contradictory equalities such as
+   :g:`n = S n`.  In this case you must use :tacn:`induction` (see
+   :ref:`example <discriminate_example>`).
+
+   If :n:`@induction_arg` is not given, the tactic checks all the hypotheses
    for impossible equalities.
-   For example, :g:`(S (S O)) = (S O)` is impossible.
    If provided, :n:`@induction_arg` is a proof of an equality, typically
    specified as the name of a hypothesis.
 
    If no :n:`@induction_arg` is provided and the goal is in the form
    :n:`@term__1 <> @term__2`, then the tactic behaves like
    :n:`intro @ident; discriminate @ident`.
```

Comment:
Your first example is `(1 = 1 -> False) -> False`, the equality is not an hypothesis, so of course it doesn't work. Maybe my sentence should be improved to make it clear that by `@term__1 <> @term__2` I mean as goal not as hypothesis
```suggestion
   If no :n:`@induction_arg` is provided , the tactic also proves goals convertible to :n:`... -> @term__1 = @term__2 -> ... -> _` with :n:`@term__1 = @term__2` an impossible equality. Note that the goal :n:`@term__1 <> @term__2` is a particular case of this as it is convertible to :n:`@term__1 = @term__2 -> False`.
```

Reply:
That's the exact same (and please avoid images when it's not needed).

Reply:
Updated.  Please take a look.  I think I covered your point of your suggestion in a more readable way.  WDYT?

Reply:
Thanks, mostly looks good, but I still think it would be worth making clear that this is just a particular case of the already described behavior (by explaining why) and not yet another exceptional behavior. In fact, this could even be an example.

Reply:
OK, I added an example for `term1 <> term2`.